src/storage-client/src/controller.rs

```diff
@@ -1785,9 +1951,109 @@ where
+        &self,
+        collection_description: &CollectionDescription<T>,
+    ) -> Vec<GlobalId> {
+        let mut result = Vec::new();
+
+        if let DataSource::Ingestion(ingestion) = &collection_description.data_source {
+            match &ingestion.desc.envelope {
+                SourceEnvelope::Debezium(envelope_debezium) => {
+                    let tx_metadata_topic = envelope_debezium.dedup.tx_metadata.clone();
+                    if let Some(tx_input) = tx_metadata_topic {
+                        result.push(tx_input.tx_metadata_global_id);
+                    }
+                }
+                // NOTE: We explicitly list envelopes instead of using a catch all to
+                // make sure that we change this when adding/removing and envelope.
+                SourceEnvelope::None(_) | SourceEnvelope::Upsert(_) | SourceEnvelope::CdcV2 => {
+                    // No storage dependencies.
+                }
+            }
+        }
+
+        result
+    }
+
+    /// Install read capabilities on the given `storage_dependencies` and return
+    /// the frontier at which we are guaranteed to be able to read from all
+    /// dependencies.
+    fn install_read_capabilities(
+        &mut self,
+        storage_dependencies: &[GlobalId],
+    ) -> Result<Antichain<T>, StorageError> {
+        // First, determine the since frontier at which we can read from all
+        // dependencies.
+        let mut dependency_since = Antichain::from_elem(T::minimum());
+        for id in storage_dependencies {
+            let collection = self.collection(*id)?;
+            let since = collection.implied_capability.clone();
+            dependency_since.join_assign(&since);
+        }
+
+        let mut changes = ChangeBatch::new();
+        for time in dependency_since.iter() {
+            changes.update(time.clone(), 1);
+        }
+
+        // Then install holds for all dependencies at the determined since.
+        let mut storage_read_updates = storage_dependencies
+            .iter()
+            .map(|id| (*id, changes.clone()))
+            .collect();
+
+        self.update_read_capabilities(&mut storage_read_updates);
+
+        Ok(dependency_since)
+    }
+
+    /// Removes read holds that were previously acquired via
+    /// `install_read_capabilities`.
+    ///
+    /// ## Panics
+    ///
+    /// This panics if there are no read capabilities at `capability` for all
+    /// depended-upon collections.
+    fn remove_read_capabilities(
+        &mut self,
+        capability: Antichain<T>,
+        storage_dependencies: &[GlobalId],
+    ) -> Result<(), StorageError> {
+        let mut changes = ChangeBatch::new();
+        for time in capability.iter() {
+            changes.update(time.clone(), -1);
+        }
+
+        // TODO: This seems involved, but we also don't do it too frequently.
+        for dep_id in storage_dependencies {
+            let dep_collection = self.collection(dep_id.clone())?;
```

Comment:
Shouldn't this check be in `update_read_capabilities`? Ensuring that we never end up with negative accumulations of timestamps?

Reply:
it's another one of the places where `update_read_capabilities` is very close to how compute does it, and it also doesn't check the updates.